File: src/transformers/features/functional/FunctionalGenericTransformer.ts
import ts from "typescript";

import { TypeFactory } from "../../../factories/TypeFactory";

import { ITransformProps } from "../../ITransformProps";
import { ITypiaContext } from "../../ITypiaContext";
import { TransformerError } from "../../TransformerError";

export namespace FunctionalGenericTransformer {
  export interface IConfig {
    equals: boolean;
  }
  export interface ISpecification {
    method: string;
    config: IConfig;
    programmer: (p: {
      context: ITypiaContext;
      modulo: ts.LeftHandSideExpression;
      expression: ts.Expression;
      declaration: ts.FunctionDeclaration;
      config: IConfig;
      init?: ts.Expression;
    }) => ts.Expression;
  }
  export const transform =
    (spec: ISpecification) =>
    (props: ITransformProps): ts.Expression => {
      // CHECK PARAMETER
      if (props.expression.arguments.length === 0)
        throw new TransformerError({
          code: `typia.functional.${spec.method}`,
          message: `no input value.`,
        });

      // GET TYPE INFO
      const type: ts.Type =
        props.expression.typeArguments && props.expression.typeArguments[0]
          ? props.context.checker.getTypeFromTypeNode(
              props.expression.typeArguments[0],
            )
          : props.context.checker.getTypeAtLocation(
              props.expression.arguments[0]!,
            );
      if (TypeFactory.isFunction(type) === false)
        throw new TransformerError({
          code: `typia.functional.${spec.method}`,
          message: `input value is not a function.`,
        });
      return spec.programmer({
        ...props,
        config: spec.config,
        expression: props.expression.arguments[0] as ts.Expression,
        declaration: type.symbol!.declarations![0] as ts.FunctionDeclaration,
        init: props.expression.arguments[1],
      });
    };
}

File: packages/typia/native/transform/features/functional/FunctionalGenericTransformer.go
package functional

import (
	shimast "github.com/microsoft/typescript-go/shim/ast"
	nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
	nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
	nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type functionalGenericTransformerNamespace struct{}

var FunctionalGenericTransformer = functionalGenericTransformerNamespace{}

type FunctionalGenericTransformer_IConfig struct {
	Equals bool
}

type FunctionalGenericTransformer_ISpecification struct {
	Method     string
	Config     FunctionalGenericTransformer_IConfig
	Programmer func(props FunctionalGenericTransformer_IProgrammerProps) *shimast.Node
}

type FunctionalGenericTransformer_IProgrammerProps struct {
	Context     nativecontext.ITypiaContext
	Modulo      *shimast.Node
	Expression  *shimast.Node
	Declaration *shimast.Node
	Config      FunctionalGenericTransformer_IConfig
	Init        *shimast.Node
}

func (functionalGenericTransformerNamespace) Transform(spec FunctionalGenericTransformer_ISpecification) func(props nativetransform.ITransformProps) *shimast.Node {
	return func(props nativetransform.ITransformProps) *shimast.Node {
		if props.Expression == nil || props.Expression.Arguments == nil || len(props.Expression.Arguments.Nodes) == 0 {
			panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
				Code:    "typia.functional." + spec.Method,
				Message: "no input value.",
			}))
		}

		var typ = props.Context.Checker.GetTypeAtLocation(props.Expression.Arguments.Nodes[0])
		if props.Expression.TypeArguments != nil && len(props.Expression.TypeArguments.Nodes) != 0 {
			typ = props.Context.Checker.GetTypeFromTypeNode(props.Expression.TypeArguments.Nodes[0])
		}
		if nativefactories.TypeFactory.IsFunction(typ) == false {
			panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
				Code:    "typia.functional." + spec.Method,
				Message: "input value is not a function.",
			}))
		}

		var declaration *shimast.Node
		if symbol := typ.Symbol(); symbol != nil && len(symbol.Declarations) != 0 {
			declaration = symbol.Declarations[0]
		}
		var init *shimast.Node
		if props.Expression.Arguments != nil && len(props.Expression.Arguments.Nodes) > 1 {
			init = props.Expression.Arguments.Nodes[1]
		}
		return spec.Programmer(FunctionalGenericTransformer_IProgrammerProps{
			Context:     props.Context,
			Modulo:      props.Modulo,
			Expression:  props.Expression.Arguments.Nodes[0],
			Declaration: declaration,
			Config:      spec.Config,
			Init:        init,
		})
	}
}
